File: Algo/hr_implement/leaderboard_denseRank.cs
using System;
using System.Linq;
using System.Collections;

public class Test
{
    public void TestFun()
    {
        int[] scores = {100, 100, 50, 40, 40, 20, 10};
        int[] alice = {5, 25, 50, 120};
        int[] result = new int[4];

        //remove duplicate and convert to array
        var y = scores.Distinct().ToArray();
        //sort ascending
        Array.Sort(y);

        int rank=0, leng = y.Length+1,i=0,mid =0;
        double high =0,low=0;

        foreach(var element in alice) 
        {
            rank = leng;
            high = leng -1;

            while(low < high)
            {
                mid = (int) Math.Round((low+high)/2);
                if(y[mid]<=element) 
                {
                    low = mid + 1;
                }
                else
                {
                    high = mid;
                }
            }

            rank -=(int)low;
            result[i] = rank;
            i++;
        }

        foreach(var element in result)
        {
            Console.WriteLine(element);
        }
    }
}

File: Algo/hr_implement/leaderboard_denseRank.js
function climbingLeaderboard(scores, alice) {
    let result = [];
    let x =  new Set(scores);
    let y = [...x];
    let len = y.length + 1,
        high = 0,
        low = 0,
        mid = 0,
        rank = 0;

    
    //descending sort
    //y.sort(function(a, b){return b-a});

    //ascending sort
    y.sort(function(a, b){return a-b});
    
    alice.forEach(element => {
        
        rank = len;
        high = len -1;
        
        while(low < high) {            
            mid = Math.floor((low+high)/2);
            if(y[mid] <= element) {             
                low = mid +1;
            } else {
                high = mid;
            }
        }
        rank -= low;
        result.push(rank);
    });
    

    return result;
}

let scores = [100, 100, 50, 40, 40, 20, 10],
    alice = [5, 25, 50, 120];

console.log(climbingLeaderboard(scores,alice));
